### PR TITLE
Adds support for typing values into single select lookups

### PR DESCRIPTION
**Pull request by mabashian**
##### SUMMARY
link #6073 

This adds support for typing values into single select lookups.  Multi-select lookups (credentials on the JT form, instance groups, etc) will remain unchanged.

An input is now rendered next to the lookup button which will allow users to type a value in.

Here's the button (which opens the modal):

Screen Shot 2021-05-25 at 9 37 52 AM

Here's the input (which lets users type a value in):

Screen Shot 2021-05-25 at 9 37 52 AM

There's a debounce on the input so that we only make a request to check to see if the name is valid after a second of no typing.

The tricky part of this implementation was handling validation on the lookup fields.  If a user types in a string that does not exactly match a value that they can use then we want to show an error indicating that no matching value was found:

Screen Shot 2021-05-25 at 9 41 25 AM

but this was tricky because some of these fields are required so the validation functions would need to change a bit.

Since the typed text and the actual underlying value of the lookup are different the validation function would need to have access to both values whenever it runs for this to work.  This meant either storing the typed text up at the form level _or_ pushing the validation down into the lookup(s).  After talking with @keithjgrant we decided to go down the route of pushing the validation down to the lookups.

You can now pass an optional validation function down to the various lookups that will get combined with the typeahead validation via the useField hook.

Here's the whole thing in action:

typeahead

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

### CODE REVIEW

**Comment by nixocio** [comment]
I triggered an e2e test run.

**Review by nixocio — APPROVED** [review]

File: awx/ui_next/src/components/Lookup/CredentialLookup.jsx [hunk 1]
```diff
@@ -111,6 +112,39 @@ function CredentialLookup({
     }
   );
 
+  const checkCredentialName = useCallback(
+    async name => {
+      if (name && name !== '') {
```
Comment (keithjgrant) [inline awx/ui_next/src/components/Lookup/CredentialLookup.jsx]: To un-nest this all a bit, you could handle the `else` case first and return early:

```
if (!name) {
  onChange(null);
  return;
}
...
```

Also, the empty string check is redundant, so probably not necessary

File: awx/ui_next/src/components/Lookup/InventoryLookup.jsx [hunk 2]
```diff
@@ -73,6 +74,24 @@ function InventoryLookup({
     }
   );
 
+  const checkInventoryName = useCallback(
+    async name => {
+      if (name && name !== '') {
```
Comment (keithjgrant) [inline awx/ui_next/src/components/Lookup/InventoryLookup.jsx]: same here with returning early to un-nest the logic one level

File: awx/ui_next/src/components/Lookup/Lookup.jsx [hunk 3]
```diff
@@ -44,9 +45,23 @@ function Lookup(props) {
...
+    fieldName,
+    validate,
   } = props;
+  const [typedText, setTypedText] = useState('');
+  const debounceRequest = useDebounce(onDebounce, 1000);
+
+  useField({
+    name: fieldName,
+    validate: val => {
+      if (!multiple && !val && typedText && typedText !== '') {
```
Comment (keithjgrant) [inline awx/ui_next/src/components/Lookup/Lookup.jsx:59]: empty string check is redundant

File: awx/ui_next/src/components/Lookup/OrganizationLookup.jsx [hunk 4]
```diff
@@ -69,6 +69,24 @@ function OrganizationLookup({
     }
   );
 
+  const checkOrganizationName = useCallback(
+    async name => {
+      if (name && name !== '') {
```
Comment (keithjgrant) [inline awx/ui_next/src/components/Lookup/OrganizationLookup.jsx]: same

File: awx/ui_next/src/components/Lookup/ProjectLookup.jsx [hunk 5]
```diff
@@ -72,6 +73,24 @@ function ProjectLookup({
     }
   );
 
+  const checkProjectName = useCallback(
+    async name => {
+      if (name && name !== '') {
```
Comment (keithjgrant) [inline awx/ui_next/src/components/Lookup/ProjectLookup.jsx]: same

**Review by keithjgrant — APPROVED** [review]
looks good! just a few minor readability suggestions

File: awx/ui_next/src/components/Lookup/CredentialLookup.jsx (hunk 1 above)
Comment (mabashian) [reply awx/ui_next/src/components/Lookup/CredentialLookup.jsx]: Good call - I like that

**Review by tiagodread — APPROVED** [review]
Verified that:
- single select lookup can be typed and find results properly
- multiple select lookup keep working as is
- all e2e tests are passing

**Pull request merged**